### ISSUE TITLE
Upgrade of Quartz .NET => Massive addition of jobs / triggers

### ISSUE THREAD

**Issue by hervesimard** [quartznet/quartznet#659]
Hello,

We have worked on an upgrade of Quartz.NET performances for adding jobs / triggers. We have a working solution which significantly boost performances. I'd like to contact you to discuss them and see if you'd like to integrate them in your solution.

Have a great day!

**Comment by lahma** [comment]
Thanks for reaching out. In fact I've been working on some storage refactoring to better support NoSQL storage engines (namely in this case RavenDB) and it would be a nice time to examine what could be done without breaking API and migration path too much.

I'd love to hear about your work, if you can add information into this issue it would be the easiest discuss and make action points out of it.

**Comment by hervesimard** [comment]
Basically we decided to give the work to the database. In our case, since we will be using Sql Server, we created stored procedure respecting the same logic and flow that was created in the code for both the jobs and triggers.
The results were impressive.
For example, for a load of about 200 000 jobs with 1 trigger each added via the `ScheduleJobs` method, it would take about 12 minutes.
Right now it takes around 1 minute 30 seconds.

**Comment by lahma** [comment]
I've investigated this batch upload and there's certainly something to improve. The code does a lot of unnecessary round-trips which can be eliminated. Unfortunately SQL Server Stored Procedure is not an universal way to handle this as Quartz needs to support different databases and database types.

**Comment by hervesimard** [comment]
Yeah, figured as much. One way would have been to do a bulk insert via SqlBulkCopy but we felt that giving the load of the work to the database was a better option since the database was actually designed for that purpose.
But we also agree with you on this point, so there's that.

**Comment by lahma** [comment]
Closing as stale — there was no follow-up after 2018 and the code has moved on since. `IScheduler.ScheduleJobs(...)` takes a dictionary of jobs to their triggers and stores them in one call (one transaction on the ADO store), and on `main` (4.0) the fire path is batched as well (#3377). If the changes you had still beat that, a pull request against `main` with numbers is welcome.
